src/pip/_internal/index/package_finder.py

```diff
@@ -452,24 +452,23 @@ def get_applicable_candidates(
         # Using None infers from the specifier instead.
         allow_prereleases = self._allow_all_prereleases or None
         specifier = self._specifier
-        versions = {
-            str(v)
-            for v in specifier.filter(
-                # We turn the version object into a str here because otherwise
-                # when we're debundled but setuptools isn't, Python will see
-                # packaging.version.Version and
-                # pkg_resources._vendor.packaging.version.Version as different
-                # types. This way we'll use a str as a common data interchange
-                # format. If we stop using the pkg_resources provided specifier
-                # and start using our own, we can drop the cast to str().
-                (str(c.version) for c in candidates),
+
+        # We turn the version object into a str here because otherwise
+        # when we're debundled but setuptools isn't, Python will see
+        # packaging.version.Version and
+        # pkg_resources._vendor.packaging.version.Version as different
+        # types. This way we'll use a str as a common data interchange
+        # format. If we stop using the pkg_resources provided specifier
+        # and start using our own, we can drop the cast to str().
+        candidates_and_versions = [(c, str(c.version)) for c in candidates]
```

Comment:
Do we know how much memory overhead this may induce in a large install? I agree this block can likely be further optimised since it is basically filtering on one list.

Reply:
> Do we know how much memory overhead this may induce in a large install? 

I ran [memray](https://github.com/bloomberg/memray) and there was no noticable memory overhead, peak memory for a dry run install of apache-airflow[all]==2.9.2 on Python 3.12 was 354 MBs, memory usage was dominated by making a list of pages of all candidates (I'm going to make a seperate issue on that).

> I agree this block can likely be further optimised since it is basically filtering on one list.

I tried making it simpler, but found that the behavior of pre-releases made it problematic. You can't filter against an individual version, because 1 pre-release will allow that pre-release, but one final  version and a pre-release will not allow that pre-release unless `allow_prereleases=True`.